packages/grafana-ui/src/components/Menu/Menu.tsx

```diff
@@ -1,23 +1,97 @@
-import React from 'react';
+import React, { useEffect, useImperativeHandle, useRef, useState } from 'react';
 import { css } from '@emotion/css';
 import { GrafanaTheme2 } from '@grafana/data';
 import { useStyles2 } from '../../themes';
+import { useEffectOnce } from 'react-use';
 
 /** @internal */
 export interface MenuProps extends React.HTMLAttributes<HTMLDivElement> {
   /** React element rendered at the top of the menu */
   header?: React.ReactNode;
   children: React.ReactNode;
   ariaLabel?: string;
+  onOpen?: (focusOnItem: (itemId: number) => void) => void;
+  onKeyDown?: React.KeyboardEventHandler;
 }
 
+const modulo = (a: number, n: number) => ((a % n) + n) % n;
+const UNFOCUSED = -1;
+type MenuItemElement = HTMLAnchorElement & HTMLButtonElement;
+
 /** @internal */
 export const Menu = React.forwardRef<HTMLDivElement, MenuProps>(
-  ({ header, children, ariaLabel, ...otherProps }, ref) => {
+  ({ header, children, ariaLabel, onOpen, onKeyDown, ...otherProps }, forwardedRef) => {
     const styles = useStyles2(getStyles);
 
+    const [focusedItem, setFocusedItem] = useState(UNFOCUSED);
+
+    const localRef = useRef<HTMLDivElement>(null);
+    useImperativeHandle(forwardedRef, () => localRef.current!);
+
+    useEffect(() => {
+      const menuItems = localRef?.current?.querySelectorAll(`[data-role="menuitem"]`);
+      (menuItems?.[focusedItem] as MenuItemElement)?.focus();
+      menuItems?.forEach((menuItem, i) => {
+        (menuItem as MenuItemElement).tabIndex = i === focusedItem ? 0 : -1;
+      });
+    }, [localRef, focusedItem]);
+
+    useEffectOnce(() => {
+      const firstMenuItem = localRef?.current?.querySelector(`[data-role="menuitem"]`) as MenuItemElement | null;
+      if (firstMenuItem) {
+        firstMenuItem.tabIndex = 0;
+      }
+      onOpen?.(setFocusedItem);
+    });
```

Comment:
is there an advantage to using this over just using `useEffect` with an empty dependency array?

Reply:
Just that our linter complains if you have an empty dependency array (which is a bit annoying tbh)